### PR TITLE
[AMLII- 2141] Add option to enforce transport protocol for log agent 

### PR DESCRIPTION
**Pull request by DDuongNguyen**
### What does this PR do?
This PR introduces a new configuration variable logs_config.transport_type that allows the Datadog Agent to explicitly force the use of the HTTP/1.1 protocol for outgoing HTTP connections. By default, the Agent uses ALPN auto-negotiation to determine the protocol (e.g., HTTP/2 or HTTP/1.1), but this new option enables users to override this behavior and enforce HTTP/1.1.

### Motivation
This change provides more control over the HTTP transport protocol, particularly for environments where HTTP/2 may not be supported or causes issues with certain proxies or servers. 

By allowing users to explicitly set the protocol, we aim to improve compatibility and provide a fail-safe option for troubleshooting network communication issues.

#### Configuration Testing:

Add `logs_config.transport_type: http1` to the Agent configuration file.
- Adjust agent logging level to trace level: `log_level: trace`.
- Confirm that log payload via HTTP connections are made using HTTP/1.1.
- Validate this by inspecting agent logs. Specifically for:
```
Log payload sent to %s. Response resolved with protocol %s in %d ms
```

#### Default Behavior Testing:
- Remove the `logs_config.transport_type` setting or set it to `auto`.
- Confirm in the logs that the Agent uses ALPN auto-negotiation for determining the protocol (typically HTTP/2).

#### Invalid Configuration Testing:
Set `logs_config.transport_type` to an invalid value (e.g., htop1).
Confirm that the Agent logs a warning and falls back to the default behavior (auto-negotiation). Log should looks like this:
`"Invalid transport_type X, falling back to 'auto'"`

### Possible Drawbacks / Trade-offs
Forcing HTTP/1.1 may reduce performance compared to HTTP/2, particularly in environments where HTTP/2 is supported and has more advantages (e.g., multiplexing, reduced latency).

### Additional Notes
- This change is backward-compatible; the default behavior remains unchanged unless the new variable is explicitly set.
- The PR includes unit tests to validate the behavior of the logs_config.transport_type
- This PR also introduce latency logging for log payload when trace log level is enabled

### CODE REVIEW

File: pkg/logs/client/http/destination.go [hunk 1]
```diff
@@ -322,6 +324,7 @@ func (d *Destination) unconditionalSend(payload *message.Payload) (err error) {
 
 	defer resp.Body.Close()
 	response, err := io.ReadAll(resp.Body)
+	log.Tracef("Log agent payload sent with: %s", resp.Proto)
```
Comment (gh123man) [inline pkg/logs/client/http/destination.go:327]: I think you can remove this log since it is mostly a dupe of the above log line.

File: pkg/logs/client/http/destination.go [hunk 2]
```diff
@@ -376,15 +379,58 @@ func (d *Destination) updateRetryState(err error, isRetrying chan bool) bool {
...
+
+	// Configure transport based on user setting
+	switch cfg.Get("logs_config.transport_type") {
+	case "http1":
+		transport.TLSNextProto = make(map[string]func(authority string, c *tls.Conn) http.RoundTripper)
+		transport.TLSClientConfig = &tls.Config{
+			NextProtos: []string{"http/1.1"},
+		}
+	case "auto":
+		// Use default ALPN auto-negotiation
```
Comment (gh123man) [inline pkg/logs/client/http/destination.go:393]: Don't we need to call `http2.ConfigureTransport` to allow the existing transport to negotiate up?

File: pkg/logs/client/http/destination.go [hunk 3]
```diff
@@ -376,15 +379,58 @@ func (d *Destination) updateRetryState(err error, isRetrying chan bool) bool {
 }
 
 func httpClientFactory(timeout time.Duration, cfg pkgconfigmodel.Reader) func() *http.Client {
+
+	transport := httputils.CreateHTTPTransport(cfg)
+
+	// Configure transport based on user setting
+	switch cfg.Get("logs_config.transport_type") {
+	case "http1":
+		transport.TLSNextProto = make(map[string]func(authority string, c *tls.Conn) http.RoundTripper)
```
Comment (gh123man) [inline pkg/logs/client/http/destination.go:388]: `http1` should not have to make changes to the transport returned from `CreateHTTPTransport` since it already forces `HTTP1.x`
These changes will override the `TLSClientConfig` previously set in `CreateHTTPTransport`

**Review by git-thuerk-done — APPROVED** [review]

File: pkg/logs/client/http/destination.go [hunk 4]
```diff
@@ -376,15 +379,58 @@ func (d *Destination) updateRetryState(err error, isRetrying chan bool) bool {
...
+			Transport: transport,
 		}
+
+		log.Tracef("Log Agent is using %v transport for connection", getTransportProtocol(client))
+		return client
 	}
 }
 
+// getTransportProtocol return the transport type
+func getTransportProtocol(client *http.Client) string {
```
Comment (gh123man) [inline pkg/logs/client/http/destination.go:412]: I don't think we know for sure what protocol is being used until we make the first request and establish a connection. 

I'm also not sure we need the log line this is used in since we will log the protocol with each payload that is posted. So IMO you can remove this.

**Review by gh123man — CHANGES_REQUESTED** [review]

File: pkg/logs/client/http/destination.go (hunk 1 above)
Comment (gh123man) [reply pkg/logs/client/http/destination.go:327]: Oh I see you are using `req.Proto` above. 
I think the above log line should be changed to `resp.Proto` as right now, it may not be accurate. + We only really care what the protocol ended up actually being after the request was made.

File: pkg/logs/client/http/destination.go (hunk 2 above)
Comment (DDuongNguyen) [reply pkg/logs/client/http/destination.go:393]: yep! I removed it without noticing

Comment (gh123man) [reply pkg/logs/client/http/destination.go:393]: > (I think http2.ConfigureTransport forces http2, but not 100% sure)
It could be worth testing this behavior. The docs are not clear.

File: pkg/logs/client/http/destination.go (hunk 1 above)
Comment (DDuongNguyen) [reply pkg/logs/client/http/destination.go:327]: Ah, actually i had a closer look at it. Seems like by default go with always send request as 1.1 but will eventually auto negotiate up to http2 if available. so logging both req and res might end up just showing mismatch (which is expected) between req.proto and resp.proto 

In that case, maybe we can just keep the response instead and use the req as logging for latency?

File: pkg/logs/client/http/destination.go [hunk 5]
```diff
@@ -376,12 +379,31 @@ func (d *Destination) updateRetryState(err error, isRetrying chan bool) bool {
...
+	case "http1":
+		// Use default ALPN auto-negotiation to negotiate up to http/1.1
+	case "auto":
+		fallthrough
+	default:
+		if cfg.Get("logs_config.transport_type") != "auto" {
+			log.Warnf("Invalid transport_type '%s', falling back to 'auto'", transport)
+		}
+		// Use default ALPN auto-negotiation and negotiate to HTTP/2 if possible, if not it will automatically fallback to best available protocol
+		http2.ConfigureTransport(transport)
```
Comment (remeh) [inline pkg/logs/client/http/destination.go:396]: You might want to catch the error, even if it's only to display a warning

**Review by remeh — CHANGES_REQUESTED** [review]

File: releasenotes/notes/force-http1-for-log-agent-2cbfba763697ab42.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+features:
+  - |
+     Introduced a new configuration variable logs_config.transport_type, allowing users to enforce HTTP/1.1 for outgoing HTTP connections in the Datadog Agent. This provides better control over transport protocols, improving compatibility with systems that do not support HTTP/2.
```
Comment (gh123man) [inline releasenotes/notes/force-http1-for-log-agent-2cbfba763697ab42.yaml:11]: Maybe we should reword this to highlight that it changes the default behavior to use HTTP2 in most cases?

File: pkg/logs/client/http/destination.go [hunk 7]
```diff
@@ -322,6 +324,7 @@ func (d *Destination) unconditionalSend(payload *message.Payload) (err error) {
 
 	defer resp.Body.Close()
 	response, err := io.ReadAll(resp.Body)
+	log.Tracef("Log agent payload resolved with: %s", resp.Proto)
```
Comment (gh123man) [inline pkg/logs/client/http/destination.go:327]: I think we can still combine this with the above log line? At line `313` we already have the `resp` instance. 
Trace logs are already very noisy so i'd suggest merging them. (essentially keep the old log line you had, but use `resp.Proto` instead of `req.proto`.

File: pkg/logs/client/http/destination.go (hunk 1 above)
Comment (gh123man) [reply pkg/logs/client/http/destination.go]: Sorry I responded in another comment. But I agree, lets keep the response proto - but merge it all into one log line for simplicity.

File: releasenotes/notes/force-http1-for-log-agent-2cbfba763697ab42.yaml (hunk 6 above)
Comment (remeh) [reply releasenotes/notes/force-http1-for-log-agent-2cbfba763697ab42.yaml:11]: But is it? Isn't the http.Client configuring for HTTP2 by default nowadays?

Comment (DDuongNguyen) [reply releasenotes/notes/force-http1-for-log-agent-2cbfba763697ab42.yaml:11]: I dove into this a bit and from what i found:

If we provide custom Dial, DialTLS, DialContext functions, or a custom TLSClientConfig (in which case we do [here](https://github.com/DataDog/datadog-agent/blob/log-agent-http/2/pkg/util/http/transport.go#L108-L122)), HTTP/2 is disabled by default **unless** configured for via:
1. External library: `http2.ConfigureTransport` [ This is better than 2nd option since it accounts for fallback to closest available protocol ]
2. Force attempt http2 via transport: `transport.ForceAttemptHTTP2 = true`

File: pkg/logs/client/http/destination_test.go [hunk 8]
```diff
@@ -366,3 +367,129 @@ func TestDestinationHA(t *testing.T) {
...
+
+	// Client send an HTTP1 request
+	resp, err := client.Do(req)
+	if err != nil {
+		t.Fatalf("Failed to make request: %v", err)
+	}
+	defer resp.Body.Close()
+
+	// Assert the protocol is HTTP/1.1
+	assert.Equal(t, "HTTP/1.1", resp.Proto)
```
Comment (carlosroman) [inline pkg/logs/client/http/destination_test.go:401]: Shouldn't this be `HTTP/1` rather than `HTTP/1.1`?  Maybe a test/option is needed to accepts `http1_1`.

File: releasenotes/notes/force-http1-for-log-agent-2cbfba763697ab42.yaml (hunk 6 above)
Comment (remeh) [reply releasenotes/notes/force-http1-for-log-agent-2cbfba763697ab42.yaml]: Oh that's interesting! Thank you

**Review by remeh — APPROVED** [review]

File: pkg/logs/client/http/destination.go [hunk 9]
```diff
@@ -344,6 +345,7 @@ func (d *Destination) unconditionalSend(payload *message.Payload) (err error) {
 		log.Debugf("Server closed or terminated the connection after serving the request with err %v", err)
 		return err
 	}
+	log.Tracef("Log payload 2 sent to %s. Response resolved with protocol %s in %d ms", d.url, resp.Proto, latency)
```
Comment (GustavoCaso) [inline pkg/logs/client/http/destination.go:348]: What does the `2` means in this context?

File: pkg/logs/client/http/destination.go [hunk 10]
```diff
@@ -393,12 +395,34 @@ func (d *Destination) updateRetryState(err error, isRetrying chan bool) bool {
 }
 
 func httpClientFactory(timeout time.Duration, cfg pkgconfigmodel.Reader) func() *http.Client {
+
```
Comment (GustavoCaso) [inline pkg/logs/client/http/destination.go:398]: Extra empty line 😄

File: pkg/logs/client/http/destination.go [hunk 11]
```diff
@@ -393,12 +395,34 @@ func (d *Destination) updateRetryState(err error, isRetrying chan bool) bool {
...
+	default:
+		if cfg.Get("logs_config.transport_type") != "auto" {
+			log.Warnf("Invalid transport_type '%v', falling back to 'auto'", transportConfig)
+		}
+		// Use default ALPN auto-negotiation and negotiate to HTTP/2 if possible, if not it will automatically fallback to best available protocol
+		err := http2.ConfigureTransport(transport)
+		if err != nil {
+			log.Warnf("Failed to configure HTTP/2 transport: %v. Resolving to best available protocol", err)
+		}
+	}
```
Comment (GustavoCaso) [inline pkg/logs/client/http/destination.go:416]: Do we see this code being used onother parts of the codebase? 

Currently, logs is the only one using it, but could there be a case that other client wants to enforce the protocol?

What about changing the `httputils.CreateHTTPTransport` function to accept an option, so we can tweak the http protocol. 

`httputils.CreateHTTPTransport(cfg, httputils.WithHttp2())`

by using the options patterrn we do not have to modify the other callers, and we can leave the http specifc code inside the `httputils ` package. 

The resulting code could look like this: 

```go
var transport *http.Transport

transportConfig := cfg.Get("logs_config.transport_type")
// Configure transport based on user setting
switch transportConfig {
case "http1":
	// Use default ALPN auto-negotiation to negotiate up to http/1.1
	transport = httputils.CreateHTTPTransport(cfg)
case "auto":
	fallthrough
default:
	if cfg.Get("logs_config.transport_type") != "auto" {
		log.Warnf("Invalid transport_type '%v', falling back to 'auto'", transportConfig)
	}
	// Use default ALPN auto-negotiation and negotiate to HTTP/2 if possible, if not it will automatically fallback to best available protocol
	transport = httputils.CreateHTTPTransport(cfg, httputils.WithHTTP2())
}
```

That way others can force HTTP2 as well and we do not have to duplicate this logic somewhere else

File: releasenotes/notes/force-http1-for-log-agent-2cbfba763697ab42.yaml [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+---
+features:
+  - |
+    Introduced a new configuration variable `logs_config.transport_type`, allowing users to enforce HTTP/1.1 for outgoing HTTP connections in the Datadog Agent. This provides better control over transport protocols and improves compatibility with systems that do not support HTTP/2.
+    By default, the log agent will now attempt to use HTTP/2 and fall back to the best available protocol if HTTP/2 is not supported.
+enhancements:
+  - |
+    Improved logging to add visiblity for latency and transport protocol
+
+
```
Comment (GustavoCaso) [inline releasenotes/notes/force-http1-for-log-agent-2cbfba763697ab42.yaml:17]: Do we need these many empty lines?

**Review by gh123man — APPROVED** [review]

File: pkg/config/config_template.yaml [hunk 13]
```diff
@@ -949,6 +949,11 @@ api_key:
   #
   # force_use_http: true
 
+  ## @param transport_type - string - optional - default: auto
+  ## @env DD_LOGS_CONFIG_TRANSPORT_TYPE - string - optional - default: auto
+  ## The transport type to use for sending logs. Possible values are "auto" or "http1".
+  # transport_type: auto
```
Comment (GustavoCaso) [inline pkg/config/config_template.yaml:955]: I'm wondering if the config key should be something `http_protocol` or `http_version` as I thin is more indicative of what we are modifying base on this config. I think `transport` is Go specific things within the http package.

**Review by GustavoCaso — APPROVED** [review]
Great work!
